service/history/executor_factory.go

```diff
@@ -0,0 +1,351 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../LICENSE -package $GOPACKAGE -source $GOFILE -destination executor_factory_mock.go
+
+package history
+
+import (
+	"context"
+
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/client"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence/visibility/manager"
+	"go.temporal.io/server/common/resource"
+	"go.temporal.io/server/common/sdk"
+	"go.temporal.io/server/common/xdc"
+	"go.temporal.io/server/service/history/archival"
+	"go.temporal.io/server/service/history/configs"
+	"go.temporal.io/server/service/history/deletemanager"
+	"go.temporal.io/server/service/history/queues"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/workflow"
+	wcache "go.temporal.io/server/service/history/workflow/cache"
+	"go.temporal.io/server/service/worker/archiver"
+	"go.uber.org/fx"
+)
+
+type (
+	ExecutorFactory interface {
+		CreateArchivalExecutor(
+			shardCtx shard.Context,
+			workflowCache wcache.Cache,
+			logger log.Logger,
+		) queues.Executor
+
+		CreateTimerExecutor(
+			shardCtx shard.Context,
+			workflowCache wcache.Cache,
+			logger log.Logger,
+			clusterName string,
+		) queues.Executor
```

Comment:
Discussed offline, but I think we should remove this and similar methods which are likely to call other methods of the same object to avoid bugs caused by the lack of virtual methods. For example: if someone overrides `CreateTimerActiveExecutor` but not `CreateTimerExecutor`, and they are surprised when their custom active timer executor isn't used.